clients/vscode-hlasmplugin/src/hlasmDownloadCommands.ts

```diff
@@ -0,0 +1,697 @@
+/*
+ * Copyright (c) 2022 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+
+import * as vscode from 'vscode';
+import { fork } from 'child_process';
+import { Client, FTPResponse, FileInfo, FTPError } from 'basic-ftp'
+import { Readable, Writable } from 'stream';
+import { EOL, homedir } from 'os';
+import path = require('node:path');
+import { promises as fsp } from "fs";
+
+export type JobId = string;
+export interface JobDescription {
+    jobname: string;
+    id: JobId;
+    details: string;
+}
+
+function getJobDetailInfo(job: JobDescription): { rc: number; spoolFiles: number } | undefined {
+    const parsed = /^.*RC=(\d+)\s+(\d+) spool file/.exec(job.details);
+    if (!parsed)
+        return undefined;
+    else
+        return { rc: +parsed[1], spoolFiles: +parsed[2] };
+}
+export interface JobClient {
+    submitJcl(jcl: string): Promise<JobId>;
+    setListMask(mask: string): Promise<void>;
+    list(): Promise<JobDescription[]>;
+    download(target: Writable | string, id: JobId, spoolFile: number): Promise<void>;
+    close(): void;
+}
+
+async function basicFtpJobClient(connection: {
+    host: string;
+    port?: number;
+    user: string;
+    password: string;
+    secure: connectionSecurityLevel
+}): Promise<JobClient> {
+    const client = new Client();
+    client.parseList = (rawList: string): FileInfo[] => {
+        return rawList.split(/\r?\n/).slice(1).filter(x => !/^\s*$/.test(x)).map((value) => new FileInfo(value));
+    };
+    await client.access({
+        host: connection.host,
+        user: connection.user,
+        password: connection.password,
+        port: connection.port,
+        secure: connection.secure !== connectionSecurityLevel.unsecure,
+        secureOptions: connection.secure === connectionSecurityLevel.unsecure ? undefined : { rejectUnauthorized: connection.secure !== connectionSecurityLevel.rejectUnauthorized }
+    });
```

Comment:
I want the function to return something that's ready to go...

Reply:
Then it might be better to replace `close` with something that does not have a pair counterpart that would feel missing in the API (`open` in this case). What about using `dispose` instead of `close`?